### PR TITLE
Variable / Rest arguments

### LINKED ISSUE #7306: [Feature Request / Lua] Type for handling Lua var args "..."
I am using Haxe to make World Of Warcraft addons and something that is used frequently is the Lua var args as a parameter "..." my proposal is to add a specific Type for this

In Haxe I would write
```haxe
eventManager.RegisterEvent("MESSAGE", (self:Frame, event:String, args:VArgs) -> {
    trace(args.length);
    if (event == "MESSAGE") {
        var msg:String = args[0];
    }
});
```
and that would compile to something like this
```lua
eventManager:RegisterEvent("MESSAGE", function(self, event, ...)
    print(select("#", ...));
    if (event == "MESSAGE") then
        local msg = select(0, ...);
    end
end);
```

Limitations would be that you can only place it in a parameter list and only as the last parameter.

### PR DESCRIPTION
**Pull request by RealyUniqueName**
Implements rest arguments as an abstract over targets native corresponding features.
Adds spread operator support for unpacking rest arguments.
Closes #7306
Softly deprecates haxe.extern.Rest (no warning for now).

- [x] js
- [x] python
- [x] php
- [x] flash
- [x] cpp
- [x] cppia
- [x] eval
- [x] lua
- [x] java
- [x] cs
- [x] jvm
- [x] hl
- [x] hl/c
- [x] neko

### CODE REVIEW

**Comment by jdonaldson** [comment]
Nice work handling the lua quirks!

**Comment by RealyUniqueName** [comment]
Anyone is willing to review?

File: src/core/ast.ml [hunk 1]
```diff
@@ -135,6 +136,7 @@ type token =
 	| Question
 	| At
 	| Dollar of string
+	| Spread
```
Comment (Aurel300) [inline src/core/ast.ml:139]: How does this differ from `Unop Spread`? Is this for parsing the function signature?

File: src/generators/genjs.ml [hunk 2]
```diff
@@ -507,9 +564,32 @@ let rec gen_call ctx e el in_value =
...
+	(match args with
+	| [_] -> ()
+	| _ -> die ~p:target.epos "`args` for `gen_call_with_apply` must contain exactly one item" __LOC__
+	);
+	match target.eexpr with
+	| TField (this, (FInstance (_,_,{ cf_name = field }) | FAnon { cf_name = field } | FDynamic field | FClosure (_,{ cf_name = field }))) ->
+		add_feature ctx "thisForCallWithRestArgs";
+		spr ctx "($_=";
+		gen_value ctx this;
+		spr ctx (",$_." ^ field ^ ".apply($_,");
```
Comment (Aurel300) [inline src/generators/genjs.ml:586]: Is this using a global `$_` to store the expression? Why not introduce a temporary variable?

File: src/core/ast.ml (hunk 1 above)
Comment (RealyUniqueName) [reply src/core/ast.ml:139]: Yes, it's because `...` can be found not only at expression level, but also in arguments lists and function signatures.

**Pull request merged**